Ssh OIDC marcus (#557)
EGI-Federation/documentation: content/en/users/compute/cloud-compute/auth/_index.md

### Using ssh-oidc

[ssh-oidc](https://github.com/EOSC-synergy/ssh-oidc) is a collection of
tools that enable ssh using OpenID-Connect access tokens.

Most of the individual tools work standalone, but they may be combined to
establish complex installations with dynamic account provisioning in

> multi-OP and multi-VO contexts.

Reviewer: what is multi-OP?

Author: Supporting more than a single OP (OIDC-Provider). I.e. it's possible to support EGI-Checkin and Indigo-IAM at the same time.
Shall I extend that description?

Reviewer 2: yes, please, not sure we are using this term anywhere in the doc, we are Check-in powered you know ;)

The tools share these common design criteria:

- None of the ssh-client or ssh-server components will be modified
- Do not store state whenever possible
    - Single exception: federated-user -> local-user mapping is stored in

>         `passwd`

Reviewer 2: ```suggestion
- Do not store state whenever possible
  - Single exception: federated-user -> local-user mapping is stored in
    `passwd`
```

- Small components that work individually (one tool for one job)

#### Installation (client-side)

`ssh-oidc` does not require specific clients per se. OIDC Access-Tokens
may simply be passed, when ssh prompts for the "Access Token".
Unfortunately, the most-popular ssh-client does not support access-tokens
which are longer than 1024 bytes _and_ EGI uses tokens that are typically
longer than 1024 bytes.

Hence, we suggest using [mccli](https://mccli.readthedocs.org) (`pip
install mccli`) on Unix/Mac. Windows users should use the putty
`oidc-plugin`, which conveniently ships with
[oidc-agent-for-windows](http://repo.data.kit.edu/windows/oidc-agent/).

For conveniently obtaining access tokens,
[oidc-agent](https://indigo-dc.gitbook.io/oidc-agent) is probably the most

> convenient choice.

Reviewer: conveniently and convenient in the same sentence?

Author: Well, oidc-agent is super convenient :)
I'll change that with the above comment.

#### Installation (server-side)

Within the federated cloud, there are multiple options for installing
the ssh-oidc server-side components:

1. Via the [Infrastructure Manager (IM)](https://im.egi.eu):
   - Choosing the `virtual machine with extra HD`.
   - Add **motley-cue** to enable SSH-OIDC.
   - Configure whether you want to allow other members of your VOs (or a
       specific VO) that should be authorised to use ssh.

   The started VM will then allow ssh access with an OIDC Access Token.

   The user that started the VM will be mapped to `cloudadm` and be able to execute
   `sudo`.

1. Manual installation:
    - Install `motley-cue`, and `pam-ssh-oidc` (or
      `pam-ssh-oidc-autoconfig`). Details in the [ssh-oid repository](https://github.com/EOSC-synergy/ssh-oidc/blob/master/installation.md).

>     - Configure `motley-cue`. See example in the [ssh-oidc repository]https://github.com/EOSC-synergy/ssh-oidc/blob/master/configuration.md).

Reviewer 2: ```suggestion
    - Configure `motley-cue`. See example in the [ssh-oidc repository](https://github.com/EOSC-synergy/ssh-oidc/blob/master/configuration.md).
```

### Using ssh-oidc

[ssh-oidc](https://github.com/EOSC-synergy/ssh-oidc) is a collection of

> tools that provide enable ssh, using OpenID-Connect access tokens.

Reviewer 3: ```suggestion
tools that enable ssh using OpenID-Connect access tokens.
```

Most of the individual tools work standalone, but they may be combined to
establish complex installations with dynamic account provisioning in
multi-OP and multi-VO contexts.

The tools share these common design criteria:

- None of the ssh-client or ssh-server components will be modified
- Do not store state whenever possible
  - Single exception: federated-user -> local-user mapping is stored in `passwd`
- Small components that work individually (one tool for one job)

#### Installation (client-side)

`ssh-oidc` does not require specific clients per se. OIDC Access-Tokens
may simply be passed, when ssh prompts for the "Access Token".
Unfortunately, the most-popular ssh-client does not support access-tokens

> which are longer than 1024 byte _and_ EGI uses tokens that are typically

Reviewer 3: ```suggestion
which are longer than 1024 bytes _and_ EGI uses tokens that are typically
```

> longer than 1024 byte.

Reviewer 3: ```suggestion
longer than 1024 bytes.
```

Hence, we suggest using [mccli](https://mccli.readthedocs.org) (`pip
install mccli`) on Unix/Mac. Windows users should use the putty
`oidc-plugin`, which conveniently ships with

> [oidc-agent-for-windows](http://repo.data.kit.edu/windows/oidc-agent/)

Reviewer 3: ```suggestion
[oidc-agent-for-windows](http://repo.data.kit.edu/windows/oidc-agent/).
```

> For conveniently obtaining access tokens,

Reviewer 3: ```suggestion
For obtaining access tokens,
```

[oidc-agent](https://indigo-dc.gitbook.io/oidc-agent) is probably the most
convenient choice.

#### Installation (server-side)

Within the federated cloud, there are multiple options for installing

> the ssh-oidc server-side components

Reviewer 3: ```suggestion
the ssh-oidc server-side components:
```

> 1. Via the Infrastructure Manager (IM) <https://im.egi.eu>: IM offers
>    starting a variety of VM infrastructures.

Reviewer 3: ```suggestion
1. Via the [Infrastructure Manager (IM)](https://im.egi.eu):
```

>    - Choosing the "virtual machine with extra HD".

Reviewer 3: ```suggestion
   - Choosing the `virtual machine with extra HD`.
```

   - Add **motley-cue** to enable SSH-OIDC.
   - Configure whether you want to allow other members of your VOs (or a
       specific VO) that should be authorised to use ssh.

   The started VM will then allow ssh access with an OIDC Access Token.

   The user that started the VM will be mapped to `cloudadm` and be able to execute
   `sudo`.

1. Manual installation:
    - Install `motley-cue`, and `pam-ssh-oidc` (or

>       `pam-ssh-oidc-autoconfig`). Details here:
>           <https://github.com/EOSC-synergy/ssh-oidc/blob/master/installation.md>

Reviewer 3: ```suggestion
      `pam-ssh-oidc-autoconfig`). Details in the [ssh-oid repository](https://github.com/EOSC-synergy/ssh-oidc/blob/master/installation.md).
```

>     - Configure `motley-cue`. Details here:
>         <https://github.com/EOSC-synergy/ssh-oidc/blob/master/configuration.md>
> 

Reviewer 3: ```suggestion
    - Configure `motley-cue`. See example in the [ssh-oidc repository]https://github.com/EOSC-synergy/ssh-oidc/blob/master/configuration.md).
```
